File: web/dart/blocks/block-style.dart
/*
 * NetTango
 * Copyright (c) 2020 Michael S. Horn, Uri Wilensky, and Corey Brady
 *
 * Northwestern University
 * 2120 Campus Drive
 * Evanston, IL 60613
 * http://tidal.northwestern.edu
 * http://ccl.northwestern.edu

 * This project was funded in part by the National Science Foundation.
 * Any opinions, findings and conclusions or recommendations expressed in this
 * material are those of the author(s) and do not necessarily reflect the views
 * of the National Science Foundation (NSF).
 */
part of NetTango;

class BlockStyle {
  static final String DEFAULT_STARTER_COLOR   = "#bb5555";
  static final String DEFAULT_CONTAINER_COLOR = "#8899aa";
  static final String DEFAULT_COMMAND_COLOR   = "#9977aa";
  static final String DEFAULT_TEXT_COLOR      = "#ffffff";
  static final String DEFAULT_BORDER_COLOR    = "#ffffff";

  String blockColor  = DEFAULT_COMMAND_COLOR;
  String textColor   = DEFAULT_TEXT_COLOR;
  String borderColor = DEFAULT_BORDER_COLOR;
  String fontWeight  = "";
  String fontSize    = "";
  String fontFace    = "";

  String get font =>
    "$fontWeight $fontSize $fontFace".trim();

  String get cssRule =>
    "color: $textColor; border-color: $borderColor; ${font == "" ? "" : "font: " + font + ";"}".trim();

  appendToSheet(CssStyleSheet sheet, String blockClass) {
    sheet.insertRule(".$blockClass-color { background-color: $blockColor; }", 0);
    sheet.insertRule(".$blockClass-attribute { color: $blockColor; }", 0);
    sheet.insertRule(".$blockClass { $cssRule }", 0);
  }

  Map toJSON() {
    return {
      blockColor:  blockColor,
      textColor:   textColor,
      borderColor: borderColor,
      fontWeight:  fontWeight,
      fontSize:    fontSize,
      fontFace:    fontFace
    };
  }

  static BlockStyle fromJSON(Map json, String blockColorDefault) {
    if (json == null) {
      return new BlockStyle() .. blockColor = blockColorDefault;
    }
    return new BlockStyle() ..
      blockColor  = toStr(json["blockColor"],  blockColorDefault)    ..
      textColor   = toStr(json["textColor"],   DEFAULT_TEXT_COLOR)   ..
      borderColor = toStr(json["borderColor"], DEFAULT_BORDER_COLOR) ..
      fontWeight  = toStr(json["fontWeight"],  "") ..
      fontSize    = toStr(json["fontSize"],    "") ..
      fontFace    = toStr(json["fontFace"],    "");
  }
}

File: src/blocks/block-style.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetLogo/NetTango

import { BlockStyle } from "../types/types"
import { ObjectUtils } from "../utils/object-utils"

class BlockStyleUI {

  // Default colors are taken from the "Pale" scheme from https://personal.sron.nl/~pault/#sec:qualitative
  // to try to have something somewhat friendly to colorblind individuals.
  // -Jeremy B June 2021
  static readonly DEFAULT_STARTER_COLOR   = "#ffcccc"
  static readonly DEFAULT_CONTAINER_COLOR = "#bbccee"
  static readonly DEFAULT_COMMAND_COLOR   = "#cceeff"
  static readonly DEFAULT_TEXT_COLOR      = "#000000"
  static readonly DEFAULT_BORDER_COLOR    = "#3b3b3b"
  static readonly DEFAULT_FONT_FAMILY     = "Poppins, sans-serif"

  static readonly DEFAULT_COMMAND_STYLE: BlockStyle = {
    blockColor:  BlockStyleUI.DEFAULT_COMMAND_COLOR
  , textColor:   BlockStyleUI.DEFAULT_TEXT_COLOR
  , borderColor: BlockStyleUI.DEFAULT_BORDER_COLOR
  , fontWeight:  ""
  , fontSize:    ""
  , fontFace:    ""
  }
  static readonly DEFAULT_CONTAINER_STYLE = ObjectUtils.clone(BlockStyleUI.DEFAULT_COMMAND_STYLE, { blockColor: BlockStyleUI.DEFAULT_CONTAINER_COLOR })
  static readonly DEFAULT_STARTER_STYLE   = ObjectUtils.clone(BlockStyleUI.DEFAULT_COMMAND_STYLE, { blockColor: BlockStyleUI.DEFAULT_STARTER_COLOR })

  readonly bs: BlockStyle

  constructor(bs: BlockStyle) {
    this.bs = bs
  }

  get font() {
    const weight = this.bs.fontWeight === "" ? "" : `font-weight: ${this.bs.fontWeight};`
    const size   = this.bs.fontSize   === "" ? "" : `font-size: ${this.bs.fontSize};`
    const face = `font-family: ${ this.bs.fontFace === "" ? BlockStyleUI.DEFAULT_FONT_FAMILY : this.bs.fontFace };`
    return `${weight} ${size} ${face}`.trim()
  }

  get cssRule() {
    return `color: ${this.bs.textColor}; border-color: ${this.bs.borderColor}; ${this.font}`.trim()
  }

  appendToSheet(sheet: CSSStyleSheet, blockClass: string) {
    sheet.insertRule(`.${blockClass}-color { background-color: ${this.bs.blockColor}; }`, 0)
    sheet.insertRule(`.${blockClass}-attribute { color: ${this.bs.blockColor}; background-color: ${this.bs.textColor}; }`, 0)
    sheet.insertRule(`.${blockClass} { ${this.cssRule} }`, 0)
  }

}

export { BlockStyleUI }
